### ISSUE TITLE
Add ability to modify scoped context right after object data is resolved

### ISSUE THREAD

**Issue by anadale** [ChilliCream/hotchocolate#742]
**Is your feature request related to a problem? Please describe.**
In order to implement resource-based authorization, I have to add custom logic **after** the object data is resolved (retrieved from the repository or anyhow resolved as part of the query processing), but **before** individual child fields of that object is resolved. The whole idea is to add custom scoped context item the moment object's data is become available (compare id of the current principal with object property and set boolean flag "current user is the owner of **this** item") so individual field directives can check that flag and report errors if the current principal is not allowed to see that field.

**Describe the solution you'd like**
Ability to add a callback/middleware/directive that allows modifying scoped context before nested fields will be resolved.

**Describe alternatives you've considered**
Passing current principal to the item resolver and add authorization information right to the data object, doesn't smell good :-(

**Comment by michaelstaib** [comment]
You can do that today already :)

so basically what you want to do is resolve the field first and then do some stuff and only after that shall the child fields be resolved/completed.

This can be done with a field middleware. You can declare those at various levels Executor/Schema/Directive/Field.

Basically a schema field middleware is applied to all the fields.


```csharp
SchemaBuilder.New()
    .AddQueryType<Foo>()
    .Use(next => async context =>
    {
        // execute the resolver pipeline
        await next(context);
        
        // do something with it
        if(context.Result is Bar)
        {
             // and maybe even change it
             context.Result = ErrorBuilder.New()
                 .SetMessage("Not Good")
                 .Build();

             // or add context data
             context.ScopedContextData = context.ScopedContextData.SetItem("foo", "bar");
        }
    })
    .Create();
```

**Comment by michaelstaib** [comment]
https://hotchocolate.io/docs/next/middleware

**Comment by michaelstaib** [comment]
The above code is V9 code ... in V8 you can do the same thing with `Schema.Create`

```csharp
Schema.Create(c => c.Use(....));
```

also if you want to only do that on one certain field you could add use via the type.


```csharp
public class QueryType
    : ObjectType<Query>
{
    protected override Configure(IObjectTypeDescriptor<Query> descriptor)
    {
        descriptor.Field(t => t.SayHello()).Type<NonNullType<StringType>>().Use(.....);
    }
}
```

Performance wise if you declare it on the field this middleware is only executed when used with the field. We compile for each field its own resolver pipeline. So, if you want it only on certain fields than declaring it in the type is the fastest way.

**Comment by michaelstaib** [comment]
Hope this is what you were looking for.

**Comment by michaelstaib** [comment]
This might also be interesting for you https://hotchocolate.io/docs/next/authorization

**Comment by anadale** [comment]
Thank you for the pointing out the possible ways to solve the problem! Two more points, though:

- Is the Field middleware chain processed in full before rendering nested fields? I mean, will calling `next` just produce the result or will also render everything down the tree?
- How to deal with the collections? Calling `next` on inside the collection field middleware will produce, for example, `List<T>`, but I need to modify scoped context for each `T`

Anyway, thank you very much for your great work, HC is really one of the rare frameworks that “just works as expected”!

**Comment by michaelstaib** [comment]
> Is the Field middleware chain processed in full before rendering nested fields? I mean, will calling next just produce the result or will also render everything down the tree?

The pipeline just produces the result. After the whole pipeline has finished the value is completed. The value completion is where we add the field result to the query result tree and enqueue the child fields for execution.

**Comment by michaelstaib** [comment]
> How to deal with the collections? Calling next on inside the collection field middleware will produce, for example, List<T>, but I need to modify scoped context for each T

So, if you are in the pipeline of a collection field you basically get the list there, again, the value completion will enqueue each item of that list after the pipeline is completed.

So, up until version 9 the execution works basically like following for queries (mutations are a bit different):

1. Enqueue root fields
2. Execute root fields (pipelines are executed in parallel)
3. Complete executed values (we now enqueue all the child resolvers to be executed)
4. Execute root fields (pipelines are executed in parallel)
5. Complete executed values (we now enqueue all the child resolvers to be executed)

and so on... it is a bit more complicated but this is basically it.

We basically implement the algorithm of the spec:
https://graphql.github.io/graphql-spec/June2018/#sec-Query

With version 9.1 however we will introduce a new execution engine that will be much more powerful.
Basically we will have execution plans that will allow you to optimize the way we execute the query graph.

For instance:

```graphql
{
    foo {
       bar {
          baz
       }
       qux {
          quux
       }
   }
}
```

The above query would with V9 first execute `foo`, after then `bar` and `qux` and after that `baz` and `quux`. With schema stitching I sometimes need `baz` to retrieve `qux`, so the V9.1 engine will be able to detect that dependency and execute first the `bar` tree and then the `qux` tree. There are other things that we than can do but basically will have real execution plan where we inspect the query first and the create a plan how to ideally execute that query.

I hope that answers your question.

**Comment by michaelstaib** [comment]
Please close the issue if the question is answered.

**Comment by anadale** [comment]
> So, if you are in the pipeline of a collection field you basically get the list there, again, the value completion will enqueue each item of that list after the pipeline is completed.

But where and how to plugin middleware to process those items?

I have solved my current problem with the Descriptor middleware, although it makes same calculations for each field and may impact performance (in the next life).

P.S. Schema stitching rocks!

**Comment by michaelstaib** [comment]
> P.S. Schema stitching rocks!

There is a lot more coming in this area.

> But where and how to plug in middleware to process those items?

you can process the items in the field middleware and update the result property with the updated items. We do something like that with paging where we get a queryable in the pipeline and that add skip and takes and pass the new queryable on.

**Comment by anadale** [comment]
Ok, thank you!

**Comment by thanhngochoang** [comment]
In version hotchocolate 12, I add a custom global variable into a data context read from just a jwt token. My case I want to add default variable "UserId" when not exist token; But in the method which I implement throw the error: The specified key `UserId` does not exist on 'context.ContextData'.
```
public sealed class JwtHttpRequestInterceptor : DefaultHttpRequestInterceptor
{
    public override ValueTask OnCreateAsync(HttpContext context,
        IRequestExecutor requestExecutor, IQueryRequestBuilder requestBuilder,
        CancellationToken cancellationToken)
    {
        if (context.User?.Identities?.All(t => !t.IsAuthenticated) == true)
        {
            // Add default
            requestBuilder.SetProperty("UserId", -1);
            return base.OnCreateAsync(context, requestExecutor, requestBuilder, cancellationToken);
        }

        var userStr = context.GetUserId();
        if (string.IsNullOrWhiteSpace(userStr) || context.User?.Identity is not ClaimsIdentity iden)
            return base.OnCreateAsync(context, requestExecutor, requestBuilder, cancellationToken);
       
        var userId = Convert.ToInt64(userStr);
        requestBuilder.SetProperty("UserId", userId);

        return base.OnCreateAsync(context, requestExecutor, requestBuilder, cancellationToken);
    }
}
```

using : 
```
    //[Authorize] remove to pass anonymous
    public long? UpdateUser([GlobalState("UserId")] long? userId)
    {
        return userId;
    }
```
pls expland me incorrect?
